### DISCUSSION TITLE
Issue with pcl::ConcaveHull while integrating PCL in VCPkg

### DISCUSSION THREAD

**Discussion by drathin4** [microsoft/vcpkg#27573] [Q&A]
I'm in the process of integrating a custom library in `vcpkg` and I'm using `vcpkg-registry` for that. I've already established communication with the support team previously in that regard:

[https://github.com/microsoft/vcpkg/discussions/27308](https://github.com/microsoft/vcpkg/discussions/url)

Before integrating this library in `vcpkg`, I'm trying to build it through `cmake` by having all the dependencies set using `vcpkg` (in `CMakeLists.txt`). 1 of the dependencies of this library is `pcl`. The build process is having issues either with `compilation` or `linking` depending on my `CMakeLists.txt` script. I'll mention both the scenarios:

There is a `ni_ptdocument.cpp` file which has `#include <pcl\surface\concave_hull.h>` and uses `pcl::ConcaveHull<pcl::PointXYZ> chull;`

CMakeLists.txt:
`find_package(PCL 1.3 REQUIRED COMPONENTS common io filters segmentation surface)`
Build Log (**compilation error**):
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1462,26): error C2039: 'ConcaveHull': is not a member of 'pcl' [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:/dev/vcpkg/packages/pcl_x64-windows/include\pcl/common/impl/transforms.hpp(58): message : see declaration of 'pcl' [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1462,37): error C2065: 'ConcaveHull': undeclared identifier [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1462,43): error C2275: 'pcl::PointXYZ': illegal use of this type as an expression [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\
nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1462): message : see declaration of 'pcl::PointXYZ' [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1462,53): error C2065: 'chull': undeclared identifier [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1478,21): error C2065: 'chull': undeclared identifier [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1479,21): error C2065: 'chull': undeclared identifier [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\fleck\Src\nimbus\foehn\ni_ptdocument.cpp(1480,21): error C2065: 'chull': undeclared identifier [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\nimbus.vcxproj]

CMakeLists.txt:
`find_package(PCL 1.3 REQUIRED COMPONENTS common io filters segmentation)`
Build Log (**linking error**):
nimbus.lib(ni_ptdocument.obj) : error LNK2019: unresolved external symbol "public: void __cdecl pcl::ConcaveHull<struct pcl::PointXYZ>::reconstruct(class pcl::PointCloud<struct pcl::PointXYZ> &,class std::vecto
r<struct pcl::Vertices,class std::allocator<struct pcl::Vertices> > &)" (?reconstruct@?$ConcaveHull@UPointXYZ@pcl@@@pcl@@QEAAXAEAV?$PointCloud@UPointXYZ@pcl@@@2@AEAV?$vector@UVertices@pcl@@V?$allocator@UVertice
s@pcl@@@std@@@std@@@Z) referenced in function "public: bool __cdecl Ni_PtDocument::DoSegmentation(void)" (?DoSegmentation@Ni_PtDocument@@QEAA_NXZ) [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\fleck.vcxproj]
nimbus.lib(ni_ptdocument.obj) : error LNK2001: unresolved external symbol "protected: virtual void __cdecl pcl::ConcaveHull<struct pcl::PointXYZ>::performReconstruction(struct pcl::PolygonMesh &)" (?performReco
nstruction@?$ConcaveHull@UPointXYZ@pcl@@@pcl@@MEAAXAEAUPolygonMesh@2@@Z) [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\fleck.vcxproj]
nimbus.lib(ni_ptdocument.obj) : error LNK2001: unresolved external symbol "protected: virtual void __cdecl pcl::ConcaveHull<struct pcl::PointXYZ>::performReconstruction(class std::vector<struct pcl::Vertices,cl
ass std::allocator<struct pcl::Vertices> > &)" (?performReconstruction@?$ConcaveHull@UPointXYZ@pcl@@@pcl@@MEAAXAEAV?$vector@UVertices@pcl@@V?$allocator@UVertices@pcl@@@std@@@std@@@Z) [D:\OneDrive - Neilsoft\Sca
n2BIM\finitor\build\fleck.vcxproj]
D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\bin\Release\fleck.exe : fatal error LNK1120: 3 unresolved externals [D:\OneDrive - Neilsoft\Scan2BIM\finitor\build\fleck.vcxproj]

I found some `vcpkg issues` related to this on the internet, but couldn't get any solution. I have already tried a few things including `set(HAVE_QHULL ON)` in `CMakeLists.txt` or `#undef DISABLE_QHULL` in the `ni_ptdocument.cpp` file, but it doesn't work. Please suggest how to resolve this issue. Thanks.

**Comment by JavierMatosD** [comment]
Hi @drathin4, are you trying to add your library to our catalog?

**Reply by JavierMatosD** [reply]
Is the version of PCL that you are getting the one fleck depends on? Would you mind giving me repro steps?

**Reply by drathin4** [reply]
The library that I'm trying to add is a proprietary library and hence not open source. I'm trying to add it to `vcpkg` only for my organization, so that it'll be available for cross platform usage across the team.

**Reply by drathin4** [reply]
As I mentioned in this thread, currently I'm just trying to build the library `fleck` using `vcpkg` in the classic mode. My next step would be to integrate `fleck` in `vcpkg` using `manifests, registries`. I'm already done writing the `vcpkg-registry`. For the sake of simplicity, I thought it would be better to build `fleck` library in a standalone manner since I had many issues while integrating it. Once this library builds successfully, I'll have the confidence to integrate it in `vcpkg`.

While building `fleck`, I was facing `compilation/linking` issues as I mentioned in the thread. Yesterday, I found a solution on `StackOverFlow`, but it involves making changes in the source code, which I don't want. Also, the solution was given more than 1.5 years ago, when there were already some issues while installing `pcl` in `vcpkg`. That discussion is present in `vcpkg issues`.
[https://stackoverflow.com/questions/66546051/pcl-not-properly-installed-with-vcpkg](url)
[https://github.com/microsoft/vcpkg/issues/17662](url)
[https://github.com/microsoft/vcpkg/issues/13881](url)

I'll attach my `CMakeLists.txt` file in here. 
[CMakeLists.txt](https://github.com/microsoft/vcpkg/files/9907573/CMakeLists.txt)

I've setup the project in a usual way and I'm using the `cmake` build commands in my project's root directory on `Windows` platform:
`mkdir build
cmake -S . -B build
cmake --build .\build --config Release
`

I want to know if I'm doing something fundamentally wrong in `CMakeLists.txt` in regards to `pcl`.

**Reply by JavierMatosD** [reply]
[CMakeLists (2).txt](https://github.com/microsoft/vcpkg/files/9913688/CMakeLists.2.txt)
I have added some comments for some minor-ish things, but I don't see anything that would be causing the linker errors. I will try to build PCL with QHULL to see if I can repro it. In the meantime, are you using the latest version of vcpkg? You also mentioned having many issues integrating with vcpkg. Would you mind describing these issues?

**Reply by drathin4** [reply]
As per your comments in `CMakeLists.txt`, I used the below statements to avoid cmake build errors.

```
set(Qt5_DIR $ENV{VCPKG_ROOT}/packages/qt5-base_x64-windows/share/cmake/Qt5)
set(PCL_DIR $ENV{VCPKG_ROOT}/packages/pcl_x64-windows/share/pcl)
```

In case I comment them out, I get the following build error:

CMake Error at D:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:829 (_find_package):
  By not providing "FindQt5.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Qt5", but
  CMake did not find one.

  Could not find a package configuration file provided by "Qt5" (requested
  version 5.9) with any of the following names:

    Qt5Config.cmake
    qt5-config.cmake

  Add the installation prefix of "Qt5" to CMAKE_PREFIX_PATH or set "Qt5_DIR"
  to a directory containing one of the above files.  If "Qt5" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:97 (find_package)

-- Configuring incomplete, errors occurred!

Additionally I have tried a few things including `set(HAVE_QHULL ON)` in `CMakeLists.txt` or `#undef DISABLE_QHULL / #define QHULL` in the `ni_ptdocument.cpp` file, but it doesn't work. Based on the `vcpkg issues` that I mentioned in the previous reply, was that `problem/bug` resolved fully in `vcpkg` or some users still face it ?

I am using the latest version of vcpkg because I installed it in September 2022 itself.
`vcpkg package management program version 2022-09-20-522aa94e9d261c7d7b2f079bf2591ca62df5c714`

The integration errors that I faced initially were all related to `CMakeLists.txt` itself. Hence I thought it would be apt to get the `CMakeLists.txt` file rectified before proceeding further. After the latest run of `vcpkg install --feature-flags=registries,manifests`, I'm getting the same linker error of `pcl::ConcaveHull`.

[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/9917533/install-x64-windows-dbg-out.log)
[install-x64-windows-dbg-out-1.log](https://github.com/microsoft/vcpkg/files/9917534/install-x64-windows-dbg-out-1.log)

**Reply by JavierMatosD** [reply]
You should never need to reach into the packages directory. Make sure you set the environment variable
```
CMAKE_TOOLCHAIN_FILE=[vcpkg root]/scripts/buildsystems/vcpkg.cmake
```
 Once that is set, ``` find_package() ``` calls should work. You might have to set the environment variable in the terminal first. 

That said, I was able to repro the bug you are experiencing with PCL. I am currently looking into it and will keep you updated. Thank you for your patience while we work this out.

**Reply by drathin4** [reply]
Thanks for your response. For the time being, I'm excluding the dependency of `PCL` from the custom library, so that the `pcl::ConcaveHull` linker error doesn't occur. But in the future, I will need it.

**Reply by JavierMatosD** [reply]
I believe I figured out the issue and opened a PR with a patch that should fix it. Here is a link to the PR with a brief description of the problem: https://github.com/microsoft/vcpkg/pull/27648. 

To confirm the patch solves your problem:

- Get my changes.
```
git fetch https://github.com/JavierMatosD/vcpkg PCL_set_qhull_found
git checkout FETCH_HEAD
```
- Remove your current copy of PCL.
```
vcpkg remove pcl
```
- Re-install it.
```
vcpkg install pcl:x64-windows
```
Please let me know if it solves your problem. :)

**Reply by JavierMatosD** [reply]
Hi @drathin4, the fix for the PCL port bug has been merged :) 
Please let me know if you are still experiencing the same problem.

**Reply by drathin4** [reply]
This fix will be useful for me in the near future, but not immediately. Thanks a lot for your help !
